Create new `str` subclass implementing postgresql bitstring for PG drivers

ovangle opened the pull request:
- Added custom `BitString` implementation which inherits from `str` and overrides methods and operators appropriately and implements the bitwise operations in a manner consistent with postgresql. Also exposes class and instance methods ensuring instances can be converted to/from `int` and `bytes` sensibly.

- Added default implementations of `bind_processor` and `result_processor` to `postgresql.BIT` type to convert `BitString` to/from string for PG drivers.

- Override `bind_processor` and `result_processor` in AsyncpgBit in order to convert `BitString` to/from `asynpg.BitString` in `asyncpg` driver.

- Added support for bitwise comparators to postgresql `BIT` instances

Fixes: #10556 

<!-- Provide a general summary of your proposed changes in the Title field above -->

### Description
<!-- Describe your changes in detail -->

### Checklist
<!-- go over following points. check them with an `x` if they do apply, (they turn into clickable checkboxes once the PR is submitted, so no need to do everything at once)

-->

This pull request is:

- [ ] A documentation / typographical / small typing error fix
	- Good to go, no issue or tests are needed
- [ ] A short code fix
	- please include the issue number, and create an issue if none exists, which
	  must include a complete example of the issue.  one line code fixes without an
	  issue and demonstration will not be accepted.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.   one line code fixes without tests will not be accepted.
- [x] A new feature implementation
	- please include the issue number, and create an issue if none exists, which must
	  include a complete example of how the feature would look.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.

**Have a nice day!**

ovangle committed:
10556: Create new BitString subclass for bitstrings for PG drivers

- Added custom `BitString` implementation which inherits from `str`
  and overrides methods and operators appropriately and
  implements the bitwise operations in a manner consistent with
  postgresql. Also exposes class and instance methods ensuring
  instances can be converted to/from `int` and `bytes` sensibly.

- Added default implementations of `bind_processor` and
  `result_processor` to `postgresql.BIT` type to convert `BitString`
  to/from string for PG drivers.

- Override `bind_processor` and `result_processor` in AsyncpgBit
  in order to convert `BitString` to/from `asynpg.BitString`
  in `asyncpg` driver.

- Added support for bitwise comparators to postgresql `BIT` instances

Fixes: 10556

sqla-tester reviewed (commented):
OK, this is **sqla-tester** setting up my work on behalf of **zzzeek** to try to get revision 761748582718bde596fcb46d7109a274b910d0cc of this pull request into gerrit so we can run tests and reviews and stuff

sqla-tester commented:
New Gerrit review created for change 761748582718bde596fcb46d7109a274b910d0cc: https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5886

zzzeek commented:
great this looks like a lot of work went into this

ovangle committed:
Fix errors from initial CE run

- fix mypy and linter issues
- expand BitString tests for int and bytes conversion
- Only run BIT test types on postgresql dialect
- Fix problems in `BitString.from_int` and `BitString.from_bytes`
- Add `BitString.to_bytes` method to allow specifying the byte
  length of the returned instances
- Fixed problems testing operators in BIT type tests

ovangle commented:
@zzzeek My pleasure, it's a couple of days at least. I don't have much else to do with myself at the moment, so I decided to start committing here. 

I spent most of today tidying up the tests (actually, most of the day after I fixed up the PG type tests was spent figuring out that `ColumnOperators` doesn't actually provide a `__xor__` override (is that because of `'#'` being used on the database side in eg postgres?)

CaselIT commented:
> The CI might not fail now, but obviously there'll be more to come. I'll rebase everything into a single commit before handing off.

there is no need to do that, you can keep committing here, the bot can do that

ovangle committed:
Fix mypy error

- `count` is not accepted as a keyword argument to str.replace in python < 3.13

sqla-tester reviewed (commented):
OK, this is **sqla-tester** setting up my work on behalf of **zzzeek** to try to get revision b4563b63895b86b86513c0976d31399316991d95 of this pull request into gerrit so we can run tests and reviews and stuff

sqla-tester commented:
Patchset b4563b63895b86b86513c0976d31399316991d95 added to existing Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5886

sqla-tester commented on lib/sqlalchemy/dialects/postgresql/asyncpg.py:
```diff
@@ -242,6 +243,29 @@ class AsyncpgTime(sqltypes.Time):
+        asyncpg_BitString = dialect.dbapi.asyncpg.BitString
+
+        def to_bind(value):
+            print(f"processing bound value '{value}'")
```
**Federico Caselli** (CaselIT) wrote:

remove debug prints in this file

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5886

sqla-tester commented on lib/sqlalchemy/dialects/postgresql/bitstring.py:
```diff
@@ -0,0 +1,347 @@
+
+    def __new__(cls, _value: str, _check: bool = True) -> BitString:
+        if not isinstance(_value, BitString) and (
+            _check and _value and any(c not in "01" for c in _value)
```
**Federico Caselli** (CaselIT) wrote:

we could init a frozenset at module level with `'01'` and do instead `allowed.union(value) > allowed` that seems to be faster in all cases

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5886

sqla-tester commented on lib/sqlalchemy/dialects/postgresql/bitstring.py:
```diff
@@ -0,0 +1,347 @@
+    def from_int(cls, value: int, length: int) -> BitString:
+        """
+        Returns a BitString consisting of the bits in the little-endian
+        representation of the given python int ``value``. A ``ValueError``
```
**Federico Caselli** (CaselIT) wrote:

I don't think this storage is little of big endian, meaning it's just a bit sequence in string form, it has no endianess. I would remove this part of the doc since it may cause confusion

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5886

ovangle replied:
The `int` which is being converted from is what might have a different endianness depending on platform, not the `BitString`.

However, I looked it up and the python docs do not actually specify which endian the `'b'` format literal uses, so this code *might* be platform dependent. I've removed the comment regardless.

sqla-tester commented on lib/sqlalchemy/dialects/postgresql/bitstring.py:
```diff
@@ -0,0 +1,347 @@
+
+        If the provided ``value`` can not be represented in a bit string
+        of at most ``length``, a ``ValueError`` will be raised. The bitstring
+        will be padded on the left by ``'0'`` to bits to produce a
```
**Federico Caselli** (CaselIT) wrote:

doc is truncated. I guess it should end with `string of the desired length`

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5886

sqla-tester commented on lib/sqlalchemy/dialects/postgresql/bitstring.py:
```diff
@@ -0,0 +1,347 @@
+        If omitted or None, 'char' defaults to "0"
+
+        e.g.
+            BitString('00010101000').rstrip() === BitString('10101000')
```
**Federico Caselli** (CaselIT) wrote:

instead of using `e.g.` like this, to properly format this as code with sphinx, use intead

```
e.g.::

    BitString('00010101000').rstrip() === BitString('10101000')
    ...
```

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5886

sqla-tester commented on lib/sqlalchemy/dialects/postgresql/bitstring.py:
```diff
@@ -0,0 +1,347 @@
+        bs: list[int] = []
+        while s:
+            bs.insert(0, int(s[-8:], 2))
+            s = s[:-8]
```
**Federico Caselli** (CaselIT) wrote:

can't we more simply do?
```py
as_int = int(self)
return as_int.to_bytes(length if length >= 0 else self.octet_length)
```

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5886

ovangle replied:
That is indeed much nicer.

sqla-tester commented on lib/sqlalchemy/dialects/postgresql/bitstring.py:
```diff
@@ -0,0 +1,347 @@
+    def __hash__(self) -> int:
+        return hash(BitString) ^ super().__hash__()
+
+    def __getitem__(self, key: SupportsIndex | slice[Any, Any, Any]) -> str:
```
**Federico Caselli** (CaselIT) wrote:

should return BitString

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5886

sqla-tester commented on lib/sqlalchemy/dialects/postgresql/bitstring.py:
```diff
@@ -0,0 +1,347 @@
+        """Return self + o"""
+        if not isinstance(o, str):
+            raise TypeError(
+                ("Can only concatenate str (not '{0}') to BitString").format(
```
**Federico Caselli** (CaselIT) wrote:

we can use f-strings

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5886

sqla-tester commented on lib/sqlalchemy/dialects/postgresql/bitstring.py:
```diff
@@ -0,0 +1,347 @@
+    def __radd__(self, o: str) -> BitString:
+        if not isinstance(o, str):
+            raise TypeError(
+                (f"Can only concatenate str (not '{0}') to BitString").format(
```
**Federico Caselli** (CaselIT) wrote:

here to, simply use f-string instead of f-string then format

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5886

sqla-tester commented on lib/sqlalchemy/dialects/postgresql/types.py:
```diff
@@ -270,6 +275,58 @@ def __init__(
+        def __lshift__(self, other: Any) -> ColumnOperators:
+            return self.bitwise_lshift(other)
+
+        def __rshift__(self, other: Any) -> ColumnOperators:
```
**Federico Caselli** (CaselIT) wrote:

would it make sense to also add the reversed operator overload for the shifts?

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5886

ovangle replied:
No, I don't think that makes sense.

     1 >> BitString('0101')
     
would mean "right shift 1 by '0101'".

The other reversed operator overloads do make sense because the operators are commutative. `'0101' | BitString('0101') == BitString('0101') | '0101'`.

I've added the missing operators.

sqla-tester commented on lib/sqlalchemy/dialects/postgresql/types.py:
```diff
@@ -270,6 +275,58 @@ def __init__(
+
+        # __xor__ is not defined on sql.operators.ColumnOperators.
+        # Use `bitwise_xor` directly instead.
+        # def __xor__(self, other: Any) -> ColumnOperators:
```
**Federico Caselli** (CaselIT) wrote:

We could add support for this in the column operators. Thoughts @zzzeek?

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5886

sqla-tester commented on test/dialect/postgresql/test_types.py:
```diff
@@ -3568,6 +3582,40 @@ def test_network_address_round_trip(
+        t.create(connection)
+
+        connection.execute(t.insert(), {"name": "test", "value": value})
+        print("value type affinity", t.c.value.type._type_affinity)
```
**Federico Caselli** (CaselIT) wrote:

remove print

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5886

sqla-tester commented on test/dialect/postgresql/test_types.py:
```diff
@@ -4169,6 +4217,112 @@ def __init__(self, name, data):
 
+class BitTests(fixtures.TestBase):
+    __dialect__ = "postgresql"
+    __only_on__ = "postgresql"
```
**Federico Caselli** (CaselIT) wrote:

should also have `__backend__=True` to test on all drivers

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5886

sqla-tester commented on test/dialect/postgresql/test_types.py:
```diff
@@ -4169,6 +4217,112 @@ def __init__(self, name, data):
+        # and it isn't a from clause of the select,
+        # the compiler doesn't actually add the column to the select's
+        # result_columns and thus the type's result_processor never gets
+        # called.
```
**Federico Caselli** (CaselIT) wrote:

this looks like a bug that should be fixed, cc @zzzeek

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5886

sqla-tester reviewed (commented):
**Federico Caselli** (CaselIT) wrote:

Great work. I've left a few comments.

This also needs a changelog entry, and likely adding of the new type in the docs here https://github.com/sqlalchemy/sqlalchemy/blob/main/doc/build/dialects/postgresql.rst?plain=1#L450

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5886

CaselIT commented:
@ovangle feel free to keep committing here, once done me or mike can ask the bot to update the gerrit and it will rebase everything in a single commit

ovangle committed:
Code review fixes

- Removed dprint statements
- Added right-hand versions of commutative operators
- Improved docstrings
- Added changelog entry
- Added dialect documentation

sqla-tester reviewed (commented):
OK, this is **sqla-tester** setting up my work on behalf of **CaselIT** to try to get revision da47f40739ff9fc6e75da44bd1663aadf80e93ca of this pull request into gerrit so we can run tests and reviews and stuff

CaselIT commented:
I'm interested in finishing this, so I'll keep it open

sqla-tester commented:
Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5886 has been **merged**. Congratulations! :)

sqla-tester commented:
**Federico Caselli** (CaselIT) wrote:

thanks for finishing this up mike

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5886
